src/core/memory.d

```diff
@@ -847,6 +847,104 @@ struct GC
     {
         gc_runFinalizers( segment );
     }
+
+    /**
+     * Queries the GC whether the current thread is running object finalization
+     * as part of a GC collection.
+     *
+     * As some GC implementations (such as the current conservative one) don't
+     * support GC memory allocation during object finalization, this function
+     * can be used to guard against such programming errors.
+     *
+     * Returns:
+     *  true if the current thread is in a finalizer, a destructor invoked by
+     *  the GC.
+     */
+    static bool inFinalizer() nothrow @nogc @safe
+    {
+        return gc_inFinalizer();
+    }
+
+    ///
+    @safe nothrow @nogc unittest
+    {
+        // Although unit tests can trigger garbage collection,
+        // and subsequent object finalization, it's highly unlikely
+        // that top-level `unittest` statements are executed from
+        // a GC finalization.
```

Comment:
If you don't call this function from a dtor there is no uncertainty here.

Reply:
I know, this what I meant, but perhaps I should have been more clear. Any suggestions on how to reword the comment?

Reply:
If a comment is necessary at all, maybe: "code not called from a dtor is never executed during finalization".